Implement analyzer for unused resources

Youssef1313 opened the pull request:
- Currently having a false positive due to https://github.com/dotnet/roslyn/issues/65969
- False positives for IVTs. The resource could be used by other projects.
- As there is no way to suppress the warning for individual cases, I think we could have an `.editorconfig` option to specify suppressions, e.g, `dotnet_code_quality.ignore_unused_resources = MyResources.Resource1,MyResources2.Resource2`

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUnusedResources.cs:
```diff
@@ -0,0 +1,81 @@
+            CreateLocalizableResourceString(nameof(AvoidUnusedResourcesTitle)),
+            CreateLocalizableResourceString(nameof(AvoidUnusedResourceMessage)),
+            DiagnosticCategory.Maintainability,
+            RuleLevel.BuildError,
```
**IMPORTANT**: I should remember to change this.

mavasani commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUnusedResources.cs:
```diff
@@ -0,0 +1,81 @@
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
Why do we need to report diagnostics on generated code?

mavasani commented:
> False positives for IVTs. The resource could be used by other projects.

I wouldn't classify these as false positives, though we may want to have an editorconfig option that bails out of resx analysis for projects that give IVT access to any assembly. IMO, if a project defines a resx that is not used within that project but used by a referencing project, then the developer should move the resx out of that project into a resx file in the referencing project.

mavasani commented:
> As there is no way to suppress the warning for individual cases, I think we could have an .editorconfig option to specify suppressions, e.g, dotnet_code_quality.ignore_unused_resources = MyResources.Resource1,MyResources2.Resource2

This seems like a good idea. I don't see any other better alternative for specifying suppressions. We may want to follow the standard pattern we use in our editorconfig options when we want to specify one or more symbol names in the option value, for example: https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#null-check-validation-methods

mavasani commented:
@Youssef1313 I understand the usefulness of this analyzer, but wanted to confirm if you opened an issue on dotnet/runtime to take this through triage. Even though the analyzer is not .NET API specific, I think it would be good to do so to ensure there are no surprises on their side when this analyzer gets added.

mavasani commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUnusedResources.cs:
```diff
@@ -0,0 +1,81 @@
+            CreateLocalizableResourceString(nameof(AvoidUnusedResourcesTitle)),
+            CreateLocalizableResourceString(nameof(AvoidUnusedResourceMessage)),
+            DiagnosticCategory.Maintainability,
+            RuleLevel.Disabled,
```
I am wondering if we should have this as `RuleLevel.Ide_Suggestion` instead of disabled by default. I understand that the rule will not report suggestion diagnostics by default as it is a compilation end analyzer, but having this as an Info diagnostic by default will mean that specifying any non-default value for [AnalysisMode](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#analysismode) will bump this to a warning and start reporting unused resource warnings on command line builds. We can also implicitly pass in all .resx files in the project as additional files when `AnalysisMode` or `AnalysisLevel` properties end with `minimum`, `recommended` or `all` similar to the below added to MS.CA.Analyzers package: https://github.com/dotnet/roslyn-analyzers/blob/79cc00251332fe690d9cbba8f367c7b21a199bc0/src/Tools/GenerateDocumentationAndConfigFiles/Program.cs#L1601-L1608

Youssef1313 commented:
Thanks for review @mavasani. I opened https://github.com/dotnet/runtime/issues/79745 and will get back to the PR once dotnet/runtime team approves the analyzer.

stephentoub commented:
This was approved in https://github.com/dotnet/runtime/issues/79765.

mavasani commented:
@Youssef1313 Do you plan to continue working on this PR?

sharwell commented:
> though we may want to have an editorconfig option that bails out of resx analysis for projects that give IVT access to any assembly

I don't see the value of an option here. Resources can't be defined as `private`, so if you want to disable analysis for an assembly with IVTs just turn off the rule completely.

Youssef1313 commented:
@mavasani Sure. I rebased and fixed the merge conflicts, updated `RuleLevel`, and updated the implementation per https://github.com/dotnet/roslyn-analyzers/pull/6338#discussion_r1050621124.

Let me know what do you think of the current implementation. Then, I think the obvious remaining task is to add an MSBuild target to the package that adds resx files as AdditionalFiles. Maybe we also need to add AdditionalFiles to CompilerVisibleItemMetadata? Not sure.

Sorry for leaving this for long.
